ruma-federation-api/src/membership/create_leave_event/v1.rs

```diff
@@ -0,0 +1,85 @@
+//! [PUT /_matrix/federation/v1/send_leave/{roomId}/{eventId}](https://matrix.org/docs/spec/server_server/r0.1.4#put-matrix-federation-v1-send-leave-roomid-eventid)
+
+use std::time::SystemTime;
+
+use js_int::UInt;
+use ruma_api::ruma_api;
+use ruma_events::{membership::MembershipEventContent, pdu::Pdu, EventType};
+use ruma_identifiers::{EventId, RoomId, ServerNameBox, UserId};
+
+ruma_api! {
+    metadata: {
+        description: "Submits a signed leave event to the resident server for it to accept it into the room's graph.",
+        name: "create_leave_event",
+        method: PUT,
+        path: "/_matrix/federation/v1/send_leave/:room_id/:event_id",
+        rate_limited: false,
+        requires_authentication: true,
+    }
+
+    #[non_exhaustive]
+    request: {
+        /// The room ID that is about to be left.
+        #[ruma_api(path)]
+        pub room_id: &'a RoomId,
+
+        /// The event ID for the leave event.
+        #[ruma_api(path)]
+        pub event_id: &'a EventId,
+
+        /// The user ID of the leaving member.
+        #[ruma_api(query)]
+        pub sender: &'a UserId,
+
+        /// The name of the leaving homeserver.
+        #[ruma_api(query)]
+        pub origin: ServerNameBox,
+
+        /// A timestamp added by the leaving homeserver.
+        #[ruma_api(query)]
+        #[serde(with = "ruma_serde::time::ms_since_unix_epoch")]
+        pub origin_server_ts: SystemTime,
+
+        /// The value `m.room.member`.
+        #[ruma_api(query)]
+        pub r#type: EventType,
+
+        /// The user ID of the leaving member.
+        #[ruma_api(query)]
+        pub state_key: String,
+
+        /// The content of the event.
+        #[ruma_api(query)]
+        pub content: MembershipEventContent,
+
+        /// This field must be present but is ignored; it may be 0.
+        #[ruma_api(query)]
+        pub depth: UInt,
+    }
+
+    #[non_exhaustive]
+    response: {
+        /// - no description -
+        #[ruma_api(body)]
+        pub event: Pdu,
```

Comment:
According to the spec, this endpoint returns `[200, {}]`, or am I missing sth.?

Reply:
No that's probably me who has been confused by @iinuwa [explanation](https://matrix.to/#/!XfqLiaavxVgyMSiRwK:matrix.org/$D8U7i4zhBCCQ2vXdAZ290RgxO_3ZAmbURhXHo4xYDw4?via=matrix.org&via=kde.org&via=feneas.org), and then get lost in the ser/de code.

More work is needed on that part...

Reply:
So... it's a Pdu but an empty one I guess... Do you want me to change the type or is it ok? (same question for v2)

Reply:
I'm pretty sure there's no such thing as an empty PDU — PDUs have a bunch of required fields.

Reply:
The description of the endpoint in the spec and response definition are confusing.

> The request and response body here describe the common event fields in more detail and may be missing other required fields for a PDU.

I think it shouldn't mention "response body" here, since there isn't a response body—or rather, it should be an empty JSON object  (`{}`) since a 200 status code requires content. (Maybe it should have been specced as a 204). The requesting homeserver is just supposed to expect a 200 status code and leave it at that.

I think the response type can just have no fields at all.

Reply:
Should I use `Empty` from `ruma-serde/src/empty.rs` then?

Reply:
Sure. I've been meaning to remove `Empty` actually, but for now this seems like the most sensible thing to do – if there was a clear name we could give this thing I'd probably prefer `struct Thing {}` over the empty type but there really isn't and it doesn't seem like there would ever be a spec revision that would add fields to this object.

Reply:
nope, in this case I think you can just do
 ```rust
    response: {}
```

@jplatte, is that right?

Reply:
We can't do that for the v1 endpoint because it has the weird `[200, <data>]` wrapping. For v2, that seems right.

Reply:
But since we're ignoring the data anyway, can't we just say that it's empty?

Reply:
We still want an endpoint implemented using this to return `[200, {}]`, not an empty response body / `{}` or anything else.

Reply:
oh right. Duh.

Reply:
I changed both endpoints reponse to `Empty` (and the v1 is wrapped with `200`).